### ISSUE TITLE
Dependency resolution changes context during packing with --no-source

### ISSUE THREAD

**Issue by muratg** [dotnet/cli#602]
_From @Tragetaschen on December 2, 2014 10:20_

I have the following setup:
```
./src/MyProject
./external/Ninject (git submodule of Tragetaschen/Ninject forked from ninject/Ninject to drop a project.json)
./external/DependencyInjection (git submodule of aspnet/DependencyInjection)
./global.json (pointing to src, external/N/src, externel/DI/src)
```
The reason is that the Nuget's Ninject does not work with Mono due to missing System.Reflection functionality and I have to compile it with the MONO flag. MyProject references Microsoft.Framework.DependencyInjection.Ninject and I'm trying get dependency resolution to pick up Ninject from the local folder.

The problem is that DependencyInjection comes with a global.json of its own and when `kpm pack --no-source` recurses into the M.F.D.Ninject, the BuildContext's _applicationHostContext reevaluates the RootDirectory and ends up with `./external/DI` instead of MyProject's RootDirectory `./`. With this RootDirectory, compilation of M.F.D.Ninject fails.This only happens with `--no-source`, copying the sources is fine.

To summarize:
- `kpm restore` in MyProject successfully finds the local Ninject and does not download the nupkg
- `kpm build` in MyProject finds (and reports) the local Ninject dependency
- `kpm pack` copies the local Ninject dependency's source code
- `kpm pack --no-source` first finds the local Ninject reference when compiling MyProject, but then fails to find it when compiling M.F.D.Ninject.

I kind of feel like the last option should succeed like packing the source does.

_Copied from original issue: aspnet/dnx#950_

**Comment by muratg** [comment]
_From @davidfowl on December 2, 2014 10:39_

Yes that sounds like a bug. Need to think about it a bit but at first glance it *should* be consistent.

**Comment by muratg** [comment]
_From @Tragetaschen on December 4, 2014 9:56_

For now, I have temporarily removed global.json and DependencyInjection.sln from my Fork of DI.

The main issue I think is: what happens when (in my example) DependencyInjection in turn also has a submodule of Ninject and references that from its global.json?
What about the following:

Instead of "sources"ing a directory, global.json also supports including other (sub-)directory's global.json files and constructs additional SourcePaths from its contents recursively. This has the added benefit, that whatever a project decides to do with its folder structure, this is automatically reflected in my build process. On the other end, the DependencyWalker/LibraryManager is shared during a full pack-compilation and thus keeps a consistent view on the environment.

In the initial example, I would include DI's global.json and add a path to Ninject. When pack builds M.F.D.Ninject, the SourcePaths include my local Ninject and it works.

In the extended example, I would include DI's global.json (which itself includes a SourcePath into its submodule) and depending on the ordering:
- Would consistently use the DI's Ninject when including the global.json *before* the path to my Ninject
- Would consistently use my local Ninject when including the global.json *after* the path to my Ninject

**Comment by muratg** [comment]
_From @troydai on December 29, 2014 23:26_

A repo: https://github.com/troydai/kruntime-issue950

**Comment by muratg** [comment]
_From @troydai on January 7, 2015 23:7_

I'll begin working on this issue now. My primary goal to to address the inconsistent behaviors between `kpm pack` and `kpm pack --no-source`. As an extended goal, I'll also look into the possibility of having a complete solution to the multiple `global.json` in a submodule git repository.

**Comment by muratg** [comment]
_From @troydai on January 8, 2015 1:13_

## Notes

### The inconsistency between `pack` and `pack no source`

When a project is being packed the action to be taken are different (see [here](https://github.com/aspnet/KRuntime/blob/dev/src/Microsoft.Framework.PackageManager/Packing/PackProject.cs#L45)) By default the pack is just merely copying files, `global.json` is not participated in the procedure. While in __no source__ the `BuildManager` will participate eventually. When `BulidManager` builds a project, an `ApplicationHostContext` is created to host the build context. The context is not aware of the parent packing context but extract the root directory from the project file path (see [here](https://github.com/aspnet/KRuntime/blob/ad6076051b491beff33d6ac5896349b4bd250600/src/Microsoft.Framework.Runtime/ApplicationHostContext.cs#L29) and [here](https://github.com/aspnet/KRuntime/blob/ad6076051b491beff33d6ac5896349b4bd250600/src/Microsoft.Framework.Runtime/ProjectResolver.cs#L66)).

### Pivots of the fix

#### The Lord of the global.jsons

One `global.json` should rule them all. While it is easy to decide which global setting file is the primary one (the nearest file upon the root project.json), there are different options to deal with the downlevel `global.json` files.

__Option 1: Enable downlevel global.json__

This is what Tragetaschen suggested, that `global.json` in downlevel folders are also included in the dependency walking. The solution won't require the developer to make additional changes to the primary `global.json`. The automated process sounds neat. However it requires the building process to deal with the contexts extremely smart because it may introduce following questions/challenges:

1. What if the dependency in downlevel global setting conflict with the parent global setting, that they reference the project with same name but at different path which takes the priority? 
2. The build process needs to keep switching context whenever it walks into a new project, and in the mean time keep tracking upper levels global settings (there could be arbitrary levels of sub projects) to resolve conflicts. 

__Option 2: Ignore downlevel global.json__

This solution assumes that when a developer adopts submodule, she/he is fully aware of what is in the sub projects and how to reference these projects. If the sub projects requires additional attention in global.json, the change should be made to the primary global.json. 

It requires the developer to deal with the primary `global.json` carefully. But the owner of the project is fully in change.

The implementation of this solution should be simplier as well. 

#### Open up `BuildManager` and `ApplicationHostContext`

No matter what solution to be taken, all the build contexts should be aware of the outer packing context. That means:
1) `BuildOptions` will carry more information about external context.
2) `ApplicationHostcontext` should accept a indication to the root directory.

**Comment by muratg** [comment]
_From @Tragetaschen on January 8, 2015 8:25_

Yes, I'm pro-Option1. However, I wouldn't go as far as to make this automatic. When I add a submodule, already I have to change my own, primary global.json to include the submodule's (src-)path to the `source`-array. I think it's perfectly fine to instead "include" the submodule's global.json and add what `source`s are mentioned there. With that, I can control which of the conflicting folders takes precedence by ordering the `source`s accordingly.

**Comment by muratg** [comment]
_From @troydai on January 8, 2015 17:28_

Thanks for the feedback @Tragetaschen .

Now here's a dilemma we may face in option 1. Assuming this sample. An application has three external submodules. Among them, one of them referee to other external references, and one of them is the a different version of upper level dependency.

``` 
root
  |-src\
  |-global.json
  |-external\
    |-lib1.version1\
      |-src\
      |-global.json
    |-lib2\
      |-src\
      |-external\
        |-lib1.version2\
        |-lib5\
      |-global.json
    |-lib3\
      |-src\
```

Here's the `global.json` of `lib2`
``` JSON
{
    "sources": [
        "src",
        "external/lib1.version2",
        "external/lib5"
    ]
}
```

Now assume using following `global.json` at the root and `lib2`'s `global.json` is explicitly call out so that it should be 'linked' into the primary `global.json`. Now which version of `lib1` is exposed to the outer? 
``` JSON
{
    "sources": [
        "src",
        "external/lib1.version2/src",
        "external/lib2/global.json"
    ]
}
```

I know this use case might look like a nit-picking, but it very much highlights the difficult in implement a correct logic of __merging__ `global.json` files. There is a similar issue in `project.json` as well #819.

Option 2 may look dumb, but by picking the referenced folders in one file, it makes the solution to similar issue clear.

**Comment by muratg** [comment]
_From @Tragetaschen on January 8, 2015 18:10_

I don't see a problem that doesn't exist today. Or I missed something fundamentally :)

Including a global.json is a shorthand for expanding its source-array. So the problem with lib1 also exists today when I use the following, primary global.json.
```
{
    "sources": [
        "src",
        "external/lib1.version2/src",
        "external/lib2/src", /* manually expanded */
        "external/lib2/external/lib1.version2", /* manually expanded */
        "external/lib2/external/lib5" /* manually expanded */
    ]
}
```
With this, the answer is as far as I understand: `external/lib1.version2/src` is exposed, which is perfectly legitimate. If I move `external/lib1.version2/src` to the bottom lib2's version is exposed.

When the submodule's global.json changes at a later date and includes new dependencies that existed before in my repository, it gets interesting. But updating any dependency blindly is bound to fail…

**Comment by smbecker** [comment]
I'm not even seeing a `--no-source` option in `dotnet pack`. Am I missing something? Is it planned for RTM?

**Comment by muratg** [comment]
@smbecker This issue was copied over from DNX with which you could either deploy source based apps or compiled apps.  With dotnet, there's no option to deploy "source" apps. So that option will not be added to dotnet because it's the default.

**Comment by smbecker** [comment]
But I still see value in --no-source to exclude source from symbols packages.

**Comment by piotrpMSFT** [comment]
@smbecker dotnet-pack will soon be replaced with the native NuGet capability. 

@rrelyea FYI.
